core/vault/vault.go

```diff
@@ -0,0 +1,254 @@
+package vault
+
+import (
+	"bytes"
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/rand"
+	"crypto/sha512"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"io"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/FleekHQ/space-daemon/core/space/domain"
+	"golang.org/x/crypto/pbkdf2"
+)
+
+type vault struct {
+	vaultAPIURL     string
+	vaultSaltSecret string
+}
+
+type VaultItemType string
+
+// Vault item types
+const (
+	PrivateKeyWithMnemonic VaultItemType = "PrivateKeyWithMnemonic"
+)
+
+type VkVersion string
+
+const (
+	VkVersion1 VkVersion = "V1"
+)
+
+// AES requires key length equal to 16, 24 or 32 bytes
+const vaultKeyLength = 32
+
+type VaultItem struct {
+	ItemType VaultItemType
+	Value    string
+}
+
+type storeVaultRequest struct {
+	Vault string `json:"vault"`
+	Vsk   string `json:"vsk"`
+}
+
+type retrieveVaultRequest struct {
+	Vsk string `json:"vsk"`
+}
+
+type retrieveVaultResponse struct {
+	EncryptedVault string `json:"encryptedVault"`
+}
+
+type Vault interface {
+	Store(uuid string, passphrase string, items []VaultItem) (*storeVaultRequest, error)
+	Retrieve(uuid string, passphrase string) ([]VaultItem, error)
+}
+
+func New(vaultAPIURL string, vaultSaltSecret string) *vault {
+	return &vault{
+		vaultAPIURL:     vaultAPIURL,
+		vaultSaltSecret: vaultSaltSecret,
+	}
+}
+
+func (v *vault) Store(uuid string, passphrase string, items []VaultItem) (*storeVaultRequest, error) {
+	// Generate vault file
+	vf, err := json.Marshal(items)
+	if err != nil {
+		return nil, err
+	}
+
+	// Compute vault key
+	vk := v.computeVk(uuid, passphrase, VkVersion1)
+
+	// Encrypt vault file using vault key
+	encVf, err := encrypt(vf, vk)
+	if err != nil {
+		return nil, err
+	}
+
+	// Compute vault service key
+	vsk := v.computeVsk(vk, passphrase, VkVersion1)
+
+	// Submit encrypted file and vsk to vault service
+	storeRequest := &storeVaultRequest{
+		Vault: base64.RawStdEncoding.EncodeToString(encVf),
+		Vsk:   base64.RawStdEncoding.EncodeToString(vsk),
+	}
+	reqJSON, err := json.Marshal(storeRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := http.Post(
```

Comment:
should we use `appToken` as authorization header, so only auth users are allowed to post keys?

Reply:
Yea, adding that when testing the integrated flow.

Reply:
Added in 8587445